### DISCUSSION TITLE
Color extraction for new blockstates

### DISCUSSION THREAD

**Discussion by maxsupermanhd** [Tnze/go-mc#159] [Show and tell]
So I argued about better blockstate interfaces recently but regardless of that here is my color generator that I made.

Single thread, no optimizations were made (only those that I could think upon writing thing by itself), res.png preserved, output is a `map[int]color.RGBA64` where key is an array index to `block.StateList`.

Will use it for my web-map WebChunk to color surface of the world.

https://gist.github.com/maxsupermanhd/cd64f374fefdeb746138b112fe474dc3

**Comment by Tnze** [comment]
I didn't had looked all the code. But I can see you are using reflect here. And that's what make you annoying I think. Go's reflecting is always annoying.

Maybe, you can marshal the block.Block into NBT, and unmarshal it back to a map[string]interface{}. And hopes that will make you happy.

Btw, good job. It will be cool to post some screenshots here.

**Reply by maxsupermanhd** [reply]
I use reflect to only check if blockstate has correct fields, not much more. Encoding in nbt/json and decoding as map[string]interface{} or something like that maybe yielded better results but I don't think so since some fields are just numbers and set via constants, after serialization I will be unable to call String method on them to get a human readable value.

**Comment by maxsupermanhd** [comment]
Here is `res.png` for all blockstates:
![res](https://user-images.githubusercontent.com/20772987/161329665-cc1b3128-1998-4e12-a265-b22379e9f48f.png)

**Reply by Tnze** [reply]
What about the final rendering result? Is there a plan of calculating shadows?

**Reply by maxsupermanhd** [reply]
Not until #161 resolved. By shadows you mean light levels? If so then most likely no for now because I want to make my WebChunk to serve minecraft server so you can view world download in minecraft

**Comment by maxsupermanhd** [comment]
Actually, from what I see after applying palette to actual terrain, I do not include tints of blocks, it is a huge problem, grass, leaves, water and all that stuff is basically white and transparent because of this... Not sure how to fix it properly since my current chunk scraper does not include biome information. Also lack of flag that block is transparent not helping at all...
